pkg/imgpkg/bundle/images_refs.go

```diff
@@ -0,0 +1,188 @@
+	}
+}
+func (i *ImageRef) updateRepo(relativeToRepo string) *ImageRef {
+	i.Image = i.replaceImageRepo(relativeToRepo)
+	return i
+}
+func (i *ImageRef) replaceImageRepo(relativeToRepo string) string {
+	imgParts := strings.Split(i.Image, "@")
+	if len(imgParts) != 2 {
+		panic(fmt.Sprintf("Internal inconsistency: The provided image URL '%s' does not contain a digest", i.Image))
+	}
+	updatedRepo := relativeToRepo + "@" + imgParts[1]
+	return updatedRepo
+}
+
+type ImageRefs struct {
+	refs                 []ImageRef
+	imageLocationsConfig *ImageLocationsConfig
+	originalImagesLock   lockconfig.ImagesLock
+
+	*sync.Mutex
+}
+
+func NewEmptyImageRefs() ImageRefs {
+	return ImageRefs{
+		Mutex: &sync.Mutex{},
+	}
+}
+
+func NewImageRefs(imagesLock lockconfig.ImagesLock, imagesLockLocationConfig ImagesLockLocationsConfig) (ImageRefs, error) {
+	imageRefs := ImageRefs{
+		Mutex:              &sync.Mutex{},
+		originalImagesLock: imagesLock,
+	}
+	for _, lockImgRef := range imagesLock.Images {
+		imageRefs.AddImagesRef(ImageRef{
+			ImageRef: lockImgRef,
+			IsBundle: nil,
+		})
+	}
+
+	err := imageRefs.syncImageRefsWithLocationConfig(imagesLockLocationConfig)
+	if err != nil {
+		return ImageRefs{}, err
+	}
+	return imageRefs, nil
+}
+
+func (i *ImageRefs) LocalizeToBundle(relativeToRepo string) {
+	for _, imgRef := range i.refs {
+		imgRef.AddLocation(imgRef.replaceImageRepo(relativeToRepo))
+		i.AddImagesRef(imgRef)
+	}
+}
+
+func (i *ImageRefs) UpdateRelativeToRepo(imgRetriever ctlimg.ImagesMetadata, relativeToRepo string) (bool, error) {
+	if i.imageLocationsConfig != nil {
+		i.LocalizeToBundle(relativeToRepo)
+		return true, nil
+	}
+
+	for _, ref := range i.refs {
+		image, err := name.ParseReference(ref.updateRepo(relativeToRepo).Image)
+		if err != nil {
+			return false, err
+		}
+		_, err = imgRetriever.Digest(image)
+		if err != nil {
+			if terr, ok := err.(*transport.Error); ok {
+				if _, ok := imageNotFoundStatusCode[terr.StatusCode]; ok {
+					return false, nil
+				}
+			}
+			return false, err
+		}
+	}
+
+	i.LocalizeToBundle(relativeToRepo)
+
+	return true, nil
+}
+func (i *ImageRefs) AddImagesRef(refs ...ImageRef) {
+	i.Lock()
+	defer i.Unlock()
+
+	for _, ref := range refs {
+		found := false
+		for j, imageRef := range i.refs {
+			if imageRef.Image == ref.Image {
+				found = true
+				i.refs[j] = ref.DeepCopy()
+				break
+			}
+		}
+
+		if !found {
+			i.refs = append(i.refs, ref)
+		}
+	}
+}
+func (i *ImageRefs) Find(ref string) (ImageRef, bool) {
+	for _, imageRef := range i.refs {
+		if imageRef.Image == ref {
+			return imageRef, true
+		}
+	}
+
+	return ImageRef{}, false
+}
+func (i *ImageRefs) MarkAsBundle(image string, isBundle bool) {
+	for j, ref := range i.refs {
+		if ref.Image == image {
+			i.refs[j].IsBundle = &isBundle
+		}
+	}
+}
+func (i ImageRefs) ImagesLock() lockconfig.ImagesLock {
+	imgLock := lockconfig.NewEmptyImagesLock()
+	for _, originalImg := range i.originalImagesLock.Images {
```

Comment:
I understand what is being asked. I don't understand the 'why'

Reply:
> we always working against a real ImagesLock

to expand on this: it seems that the intent for this class is to be able to generate ImagesLock based on _existing images lock_ + image refs. using this method against an ImageRefs instance that was never initialized with ImagesLock should be a programmer's mistake because we know that we must always be basing new images lock on existing images lock -- double-checking system invariants is generally a good pattern (unless its in the middle of the perf sensitive code). currently if this method is called on NewEmptyImagesRefs, it will just return empty ImagesLock. it's probably better to blow up to catch that or any other possibility instead of generating erroneous ImagesLock. this change ultimately firms up the contract of this class.

Reply:
Thanks for the explanation. good point i'll make the change.